anta/tests/stun.py

```diff
@@ -0,0 +1,108 @@
+    """
+
+    name = "VerifyStunClient"
+    description = "Verifies the STUN client is configured with the specified IPv4 source address and port. Validate the public IP if provided."
+    categories: ClassVar[list[str]] = ["stun"]
+    commands: ClassVar[list[AntaCommand | AntaTemplate]] = [AntaTemplate(template="show stun client translations {source_address} {port}")]
+
+    class Input(AntaTest.Input):
+        """Input model for the VerifyStunClient test."""
+
+        stun_clients: list[ClientAddress]
+
+        class ClientAddress(BaseModel):
+            """Source and public address details of STUN client."""
+
+            source_address: IPv4Address
+            """IPv4 source address of STUN client."""
+            public_address: IPv4Address | None = None
+            """IPV4 public address of STUN client."""
+            port: Port
+            """Port number for STUN client."""
+
+    def render(self, template: AntaTemplate) -> list[AntaCommand]:
+        """Render the template for each STUN translations."""
+        return [template.render(source_address=client.source_address, port=client.port) for client in self.inputs.stun_clients]
+
+    @AntaTest.anta_test
+    def test(self) -> None:
+        """Main test function for VerifyStunClient."""
+        self.result.is_success()
+
+        for command, client_input in zip(self.instance_commands, self.inputs.stun_clients):
+            bindings = command.json_output["bindings"]
+            source_address = str(command.params.source_address)
+            port = command.params.port
+            if not bindings:
+                self.result.is_failure(f"STUN client transaction for source `{source_address}:{port}` is not found.")
+                continue
+            public_address = client_input.public_address
+
+            transaction_id = next(iter(bindings.keys()))
+            actual_stun_data = {
+                "source ip": get_value(bindings, f"{transaction_id}.sourceAddress.ip"),
+                "source port": get_value(bindings, f"{transaction_id}.sourceAddress.port"),
+            }
+            expected_stun_data = {"source ip": source_address, "source port": port}
+
+            # Skip public address verification when the public address is not defined in the input
+            if public_address is None:
+                if actual_stun_data != expected_stun_data:
+                    failed_log = get_failed_logs(expected_stun_data, actual_stun_data)
+                    self.result.is_failure(f"For STUN source address {source_address}:{failed_log}")
+
+            else:
+                actual_stun_data.update(
+                    {
+                        "public ip": get_value(bindings, f"{transaction_id}.publicAddress.ip"),
+                        "public port": get_value(bindings, f"{transaction_id}.publicAddress.port"),
+                    }
+                )
+                expected_stun_data.update({"public ip": str(public_address), "public port": port})
```

Comment:
the port and public_port could be different (this is the purpose of STUN to discover your public IP and public port)

Reply:
```
edge4B#show  stun client translations
Current System Time: Thu Apr 11 08:55:36 2024
Agent           Transaction ID Source Address        Public Address        Last
                                                                           Refreshed
----- ------------------------ --------------------- --------------------- ------------
dps   00000001c0a8c80800000000 10.3.43.1:4500        10.3.43.1:6006        0:00:00 ago

edge4B#show  stun client translations json
% Invalid input
edge4B#show  stun client translations | json
{
    "bindings": {
        "00000001c0a8c80800000000": {
            "agentName": "dps",
            "sourceAddress": {
                "ip": "10.3.43.1",
                "port": 4500
            },
            "publicAddress": {
                "ip": "10.3.43.1",
                "port": 6006
            },
            "lastRefreshed": 7,
            "timeoutInterval": 180,
            "tlvs": []
        }
    }
}
```